x/mock/simulation/rand_util.go

```diff
@@ -62,3 +62,29 @@ func RandTimestamp(r *rand.Rand) time.Time {
 	unixTime := r.Int63n(253373529600)
 	return time.Unix(unixTime, 0)
 }
+
+// Derive a new rand deterministically from a rand.
+// Unlike rand.New(rand.NewSource(seed)), the result is "more random"
```

Comment:
Why is this "more random"? I don't really follow - is there a substantive difference between reinitializing a new PRNG and just stepping the same PRNG?

cc @ValarDragon

Reply:
This is not more random in a meaningful way. This doesn't even increase the repetitivity of the internal RNG state. (Since your combining multiple PRNG's with the same cycle size via xor)

This typically may be done to combine more entropy in some strange settings, however here that is also not the case since each seed is derived from the original `r` instance.

The rng used in golang is essentially a lagged fibonacci random number generator, with fixes that improve the seeding. (Source: https://groups.google.com/forum/#!topic/golang-nuts/RZ1G3_cxMcM) That has an insane cycle length (2^603) and is sufficiently random for our purposes. It is not suitable for things like monte-carlo simulations, but nor is this. It would make sense to switch to a monte-carlo safe rng in the future.

Reply:
The repetitivity/cycle-length doesn't matter, as you mentioned.  A problem of the original math.Rand (with default Source) is that `seed values that have the same remainder when divided by 2^31-1 generate the same pseudo-random sequence.`.  There are only two billion unique initial random sequences for a new Rand (with default Source)!

In order to make simulation deterministic while being able to make changes in the middle somewhere, we can't just re-use the same *Rand, because changes to the usage of the *Rand of a middle transaction causes rippling changes for all future operations.  In some situations this can make debugging more difficult, because e.g. just because you mutate/mute/remove a single operation of a block doesn't mean that you're really testing the same scenario, as later operations of the same block (as an example) would be completely different.

The solution is to construct a tree of rands, as it's already done here... Here, we construct a predetermined set of operations, and then for further logic within an operation, we pass in a derived Rand.  In the future we may want to derive third and fourth order rands in this way.

Once you're deriving a bunch of rands in this way, the limitation of 2^31-1 possible sequences does become meaningful.  Sometimes you'll end up with the same seed for a derived rand, which means for that branch of rand usage, it's no longer random... the same sequence was used in another branch of simulation logic!

This DeriveRand() function is designed to solve this issue.  The possibility of deriving the same rand sequence becomes negligible.

Reply:
Thanks for the clarification, I didn't know that! The comment should be amended from making the result more "random", which its not, to instead increasing the seed space.

However I still don't think this is worth doing. With there being 2**31 different valid seeds, we'd have to do 54 thousand simulations with no change to the simulator in order for there to be a 50% probability of a single duplicate seed. We'd have to do 8.3 million simulations before there is a 1% probability each successive simulation was for nought. I frankly don't think we will do 54 thousand long simulations ever before changing the simulator, let alone 8.3 million. I don't think this is worth the performance trade-off, as speed of the simulator matters.